### ISSUE TITLE
tribits/cmake scripts:  no support for AMD libraries

### ISSUE THREAD

**Issue by kliegeois** [trilinos/Trilinos#11359]
## Bug Report
@trilinos/tribits @bartlettroscoe 

### Description
There is currently no support at the Trilinos level to support AMD TPLs such as rocblas or rocsparse.

This leads to compilation/linking issues in at least two cases:

- When compiling tacho when Kokkos_ENABLE_HIP is true
- When compiling Kokkos Kernels when Trilinos is configured with:
```-DKokkosKernels_ENABLE_TPL_ROCSPARSE=ON``` 
or/and
```-DKokkosKernels_ENABLE_TPL_ROCBLAS=ON```

This issue could be fixed by handling the AMD TPLs as we handle the NVIDIA ones at the Trilinos level.

@lucbv

**Comment by bartlettroscoe** [comment]
This should be straightforward for Trilinos and should not need any updates to TriBITS.  See:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-add-a-new-tribits-external-package-tpl

I am wondering if this is something the @trilinos/framework team can cut their CMake/TriBITS teeth on? (CC: @sebrowne)

**Comment by kliegeois** [comment]
Sorry @bartlettroscoe was not sure which team to ping for this.

**Comment by bartlettroscoe** [comment]
> Sorry @bartlettroscoe was not sure which team to ping for this.

@kliegeois, I will let @ccober6 comment on this, but I believe that build/test system issues/requirements for Trilinos should go to the @trilinos/framework team and then they can consult me if needed.  I believe there should be documentation for 90% of the use cases for TriBITS in:

* https://tribitspub.github.io/TriBITS/users_guide

(but it may be hard to find exactly what one needs).

CC: @mmwolf, @sebrowne

**Comment by ccober6** [comment]
@kliegeois we are trying to setup the process of the Framework handling/triaging issues initially.  If additional expertise is needed by the Framework (e.g., @bartlettroscoe), then they can bring them in as needed.  This allows the Framework to be aware of the issue, fix it if they have the expertise, and/or learn from the experts on the appropriate changes.

As always, we do not want to short circuit useful discussions between developers (e.g., yourself and @bartlettroscoe), we are just trying to setup a process that includes the right people at the right time.

CC: @mmwolf @sebrowne

**Comment by sebrowne** [comment]
Is this effectively just adding the ROCm TPLs to Trilinos as they've apparently been added to Kokkos?

**Comment by bartlettroscoe** [comment]
> Is this effectively just adding the ROCm TPLs to Trilinos as they've apparently been added to Kokkos?

@sebrowne, that is what I suspect.  The TriBITS External Package/TPL approach is very flexible, especially after the upgrade to use Modern CMake imported targets.

**Comment by jjellio** [comment]
ROCM is a mess for adding the libraries. IMO, a correct approach for Trilinos/Tribits, would be a a top-level ROCm enable, and then test/try to find all the various rocm comments - report what is findable, and enable that. There are a few ROCm packages that matter, and unfortunately they aren't always installed - and right now, even if they are, the locations are not standardized.

This also could be messy if Kokkos or KK try to find ROCm.

**Comment by kliegeois** [comment]
@jjellio currently, Kokkos Kernels does not try to find ROCblas or ROCsparse when compiled inside Trilinos.
Kokkos Kernels looks for them when compiled as a stand-alone.

**Comment by lucbv** [comment]
@jjellio the approach you are proposing seem to imply that the user should enable ROCm (or HIP whatever we call it) and hope the TPLs are in standard places and get detected.
On the other you state that these TPLs are not always in the standard place and using the approach you outline that would ban the user from telling Trilinos directly where TPLs are when they are not installed in a standard place which would basically forbid there usage. I am guessing this is not what you intend though?

**Comment by jjellio** [comment]
Sorry, I didn't see this mention.

I tend to think, if HIP is chosen as an enabled Kokkos Component, then someone (Trilinos?) could just enumerate all possible HIP things, and then write down: "We will try to find BLAS, SPARSE, _____"

This would be as simple as

```
ENABLING HIP Support:
Detecting available Components:
- Finding ROCBlas: 
- Finding HIPBlas:
- Finding ROCSparse:
- Finding HIPSparse:
- Finding ConMgr:
- ...
```

Whether CMake can "find" something without enable it... is a different question. Reporting what is found/what is not would be valuable. IMO, if rocm components are installed it seems they should be enabled by default (since they tend to intra-depend).

The list of ROCm/HIP things is quite long, but we could always try to push this back to CMake as some type of meta-ROCM module. (In the spirit that NetCDF tries to do reasonable HDF5 detection)

**Comment by fnrizzi** [comment]
after discussing with @lucbv , @mperrinel and I will try to do something for this and see what solution we can come up with.

**Comment by jhux2** [comment]
@fnrizzi @lucbv Is there any update on this?

**Comment by lucbv** [comment]
Yes, this was actually done with PR #11750 just above your comment, we can probably just close this

**Comment by jjellio** [comment]
Based on that PR, it seems Trilinos people need to `TPL_ENABLE_ROCBLAS and TPL_ENABLE_ROCSPARSE` - is that correct? That's a deviation from how Cuda was done.

**Comment by bartlettroscoe** [comment]
> Based on that PR, it seems Trilinos people need to `TPL_ENABLE_ROCBLAS and TPL_ENABLE_ROCSPARSE` - is that correct? That's a deviation from how Cuda was done.

@jjellio, actually, Trilinos does define `CUBLAS` and `CUSPARSE` TPLs.  See:

https://github.com/trilinos/Trilinos/blob/83ed4dee9ac65858169c9ecfefa02e0a873f58c7/TPLsList.cmake#L61-L63

and:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/TPLs/FindTPLCUBLAS.cmake
* https://github.com/trilinos/Trilinos/blob/develop/cmake/TPLs/FindTPLCUSPARSE.cmake

**Comment by jjellio** [comment]
That's true Ross, but the semantics wtih Cuda (enabled inside Tpetra) - not as a TPL, was that cuBLAS was enabled silently by default.

This is *different* from enabling a TPL.  The TPL says "A downstream users intends to develop against those packages" - what I'm talking about here is that the user expects the internal bits of Trilinos to use those libs. With CUDA, you never needed to explicitly enable any TPLs actually. Enabling the Kokkos execution space triggered Tpetra's use of it, which implciitly used cublas - CuSparse did require the TPL to be enabled - which if *true*, then packages like MueLu would allow it to be a data type used.

So there are two distinct issues - one is providing a TPL so people can see the API (get the headers) and link against it sanely. But the broader and most important issue is what KokkosKernels does by default.

**Comment by lucbv** [comment]
@jjellio what you say is correct but ignores the fact that NVIDIA is packaging it's libraries differently from AMD which is really at the source of some of these differences. Basically cuda always installed cublas and cusparse (although nvhpc and cudatoolkit are potentially doing things differently). On the other hand I do not believe that rocm makes the same promises so we cannot assume that we know if/where rocblas and rocsparse are installed

**Comment by jjellio** [comment]
I, of course, disagree with that entirely. On the large scale HPC deployments from ORNL and LLNL I can promise you the full AMD stack will be installed - it's written in the procurement.  My gripe is that users rarely have a clue what low level kernels actually get used - and the only way to find out is via profiling, which is obnoxious.

I have no issue with Kokkos kernels - it serves a *needed* purpose of providing a dispatch layer and it offers the much needed ability to execute BLAS/LAPACK (among others) - on non-standards datatypes (that is essential for UQ).  But the default behavior that every install doesn't bother to try finding AMD's tools "because they aren't where CUDA's are" is a weak excuse in my book - and atleast should be coupled with an effort + warning - "Hey we couldn't find the vendor's tuned libraries so you aren't using them".  Or better, a way for teams to express which default they want. 

In the past, it's been that vendor TPLs for BLAS and Sparse have excelled in some functions and stunk in others (but that was mostly Sparse - and mostly 5+ years back.). Being able to specify, "I want TPL DOT with Kokkos GEMM" would be useful, but would require a fairly complicated build system.

In my view, the user should default to getting Vendor TPLs for BLAS/LAPACK and need to request KK (not the other way around). Or better, just have device BLAS/LAPACK/SPARSE (lapack isn't really a thing on the device that I've seen, but BLAS and CRS certainly is mature)

**Comment by lucbv** [comment]
@jjellio if you want to have a quick chat about this I'm happy to set up a discussion with you and we can find a reasonable approach. My only requirements are the following:

1. we should have a more uniform approach
2. it should provide reasonable support to all customers
3. Trilinos needs to agree to it , needs to support it and cannot change their mind too frequently...

Let me know if that sounds reasonable?

**Comment by bartlettroscoe** [comment]
One thing that I should note (that I failed to realize from the review of PR #11750) is that once a TPL is enabled, it must be found or that is an error.  The problem with the current implementation from PR #11750 is that if `find_package(ROCBLAS)` or `find_package(ROCSPARSE)` don't find those packages, then you will get a configure-time error complaining about the absences of the `ROCBLAS::all_libs` and/or `ROCSPARSE::all_libs` targets.

There are one of two ways to address this.  The first way is to change [FindTPLROCBLAS.cmake](https://github.com/trilinos/Trilinos/blob/develop/cmake/TPLs/FindTPLROCBLAS.cmake) and [FindTPLROCSPARSE.cmake](https://github.com/trilinos/Trilinos/blob/develop/cmake/TPLs/FindTPLROCSPARSE.cmake) to call:

```
find_package(<ExternalPkg> REQUIRED)
```

which will yield a hard error-out with a clear error message.  (See [Creating `FindTPL<tplName>.cmake` using `find_package()` with `IMPORTED` targets](https://tribitspub.github.io/TriBITS/users_guide/index.html#creating-findtpl-tplname-cmake-using-find-package-with-imported-targets).)

The second way is to declare these to be [tentatively enabled external packages/TPLs](https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-tentatively-enable-an-external-package-tpl)) and leave their `FindTPL<ExternalPkg>.cmake` files using `find_package(<ExternalPkg>...)` as they are.  (It looks like TriBITS may need a tweak for this specific use case.)

One purpose of the TriBITS External Package/TPL system is to make a CMake package's/project's dependencies on external software very explicit and to define behavior when external packages/TPLs are enabled or disabled.

**Comment by jjellio** [comment]
@lucbv I think a really start (which really augments this PR) - would simply have CMake logic (in Trilinos) - where Cmake tries to find some predefined list of ROCM things - and reports what it finds.

@bartlettroscoe - Atleast with MPICH + AMD, there's a need to add additional MPI libraries - but as-is we can't define the libs for MPI for Trilinos. This is somewhat related, but could be a separate issue  - finding some of those libs implies finding a valid rocm to begin with - others are the 'glue' between rocm and mpich. Right now, on the early access systems it's a mess of different approaches that solve this. Some sites are providing `mpicc` wrappers, Cray provides their PE (CC,cc,ftn) wrappers, and then you have alot of us that are just tacking on all the libs ourselves - because each wrapper introduces some form of headache.

If you'd like I can share what my configure scripts can look like - to see what I am doing in some cases because I don't have a clean way to do it.

**Comment by kliegeois** [comment]
Can we close this one @jjellio ?
